lib/clientv3/ippool_kdd_conversion_test.go

```diff
@@ -0,0 +1,249 @@
+			Kind: apiv3.KindIPPool,
+		},
+		Value: &apiv3.IPPool{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       apiv3.KindIPPool,
+				APIVersion: apiv3.GroupVersionCurrent,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "2001---120",
+			},
+			Spec: apiv3.IPPoolSpec{
+				CIDR:     "2001::/120",
+				Disabled: false,
+				IPIP: &apiv1.IPIPConfiguration{
+					Enabled: false,
+				},
+			},
+		},
+	}
+	DescribeTable("IPPool CRD with v1 IPIP field tests",
+		func(name1, name2 string, spec_v1, spec_v3 apiv3.IPPoolSpec, kvp *model.KVPair) {
+			c, err := clientv3.New(config)
+			Expect(err).NotTo(HaveOccurred())
+
+			be, err := backend.NewClient(config)
+			Expect(err).NotTo(HaveOccurred())
+			be.Clean()
+
+			By("Attempting to creating a new IPPool with the non-writable v1 IPIP field")
+			_, outError := c.IPPools().Create(ctx, &apiv3.IPPool{
+				ObjectMeta: metav1.ObjectMeta{Name: name1},
+				Spec:       spec_v1,
+			}, options.SetOptions{})
+			Expect(outError).To(HaveOccurred())
+
+			By("Creating IPPool with v1 IPIP field directly in the backend")
+			_, err = be.Create(ctx, kvp)
+			Expect(err).NotTo(HaveOccurred())
+
+			outPool, err := be.Get(ctx, kvp.Key, "")
+			Expect(err).NotTo(HaveOccurred())
+			Expect(outPool.Value.(*apiv3.IPPool).Spec.IPIP).To(BeNil())
+			Expect(outPool.Value.(*apiv3.IPPool).Spec).To(Equal(spec_v3))
+
+			By("Updating the IPPool from the API client with the non-writable v1 IPIP field")
+			_, outError = c.IPPools().Update(ctx, &apiv3.IPPool{
+				ObjectMeta: metav1.ObjectMeta{Name: name1, ResourceVersion: "555", CreationTimestamp: metav1.Now(), UID: "a-rabbit-ate-my-carrot"},
+				Spec:       spec_v1,
+			}, options.SetOptions{})
+			Expect(outError).To(HaveOccurred())
+
+			By("Listing all the IPPools, expecting a single result with name1/spec_v1")
+			outList, outError := c.IPPools().List(ctx, options.ListOptions{})
+			Expect(outError).NotTo(HaveOccurred())
+			Expect(outList.Items).To(HaveLen(1))
+			testutils.ExpectResource(&outList.Items[0], apiv3.KindIPPool, testutils.ExpectNoNamespace, name1, spec_v3)
+
+			By("Creating a new IPPool with name2/spec_v3")
+
+			By("Creating another IPPool with v1 IPIP field directly in the backend name2/kvp_v3")
+			kvp.Key = model.ResourceKey{
+				Name: name2,
+				Kind: apiv3.KindIPPool,
+			}
+			_, err = be.Create(ctx, kvp)
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Listing all the IPPools, expecting a two results with name1/spec_v1 and name2/spec_v3")
+			outList, outError = c.IPPools().List(ctx, options.ListOptions{})
+			Expect(outError).NotTo(HaveOccurred())
+			Expect(outList.Items).To(HaveLen(2))
+			testutils.ExpectResource(&outList.Items[0], apiv3.KindIPPool, testutils.ExpectNoNamespace, name1, spec_v1)
+			testutils.ExpectResource(&outList.Items[1], apiv3.KindIPPool, testutils.ExpectNoNamespace, name2, spec_v3)
+
+		},
+
+		Entry("IPv4 IPPool CRD with v1 IPIP field and IPIP Mode CrossSubnet", name1, name2, spec1_v1, spec1_v3, kvp1_v3),
```

Comment:
One for IPIPEnabled and Always?
One for Disabled and Always?

Reply:
what do you mean by Disabled and Always? how can that be the case? First thing we check is if it's disabled then set the mode to Never